jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalUpcallHandler.java

```diff
@@ -0,0 +1,187 @@
+		realReturnLayout = cDesc.returnLayout().orElse(null); // Set to null for void
+
+		/*[IF JAVA_SPEC_VERSION <= 17]*/
+		/* The layout check against the method type is still required for Java 16 & 17 in that
+		 * both the function descriptor and the method type are passed in as arguments by users.
+		 * Note: skip the validity check on function descriptor in Java 18 as it is done before
+		 * initializing ProgrammableUpcallHandler in OpenJDK. Meanwhile, the method type is
+		 * directly deduced from the function descriptor itself, in which case there is no need
+		 * to check the layout against the method type.
+		 */
+		TypeLayoutCheckHelper.checkIfValidLayoutAndType(mt, argLayoutArray, realReturnLayout);
+		/*[ENDIF] JAVA_SPEC_VERSION <= 17 */
+
+		/*[IF JAVA_SPEC_VERSION >= 19]*/
+		thunkAddr = getUpcallThunkAddr(target, session);
+		/*[ELSE] JAVA_SPEC_VERSION >= 19 */
+		thunkAddr = getUpcallThunkAddr(target, scope);
+		/*[ENDIF] JAVA_SPEC_VERSION >= 19 */
+	}
+
+	/**
+	 * Returns the address of the generated thunk at runtime.
+	 *
+	 * @return the thunk address
+	 */
+	public long upcallThunkAddr() {
+		return thunkAddr;
+	}
+
+	/* Check whether the thunk address of the requested java method exists in the cache;
+	 * otherwise, call the native to request the JIT to generate an upcall thunk for this
+	 * java method.
+	 */
+	/*[IF JAVA_SPEC_VERSION >= 19]*/
+	private long getUpcallThunkAddr(MethodHandle target, MemorySession session)
+	/*[ELSE] JAVA_SPEC_VERSION >= 19 */
+	private long getUpcallThunkAddr(MethodHandle target, ResourceScope scope)
+	/*[ENDIF] JAVA_SPEC_VERSION >= 19 */
+	{
+		int argLayoutCount = argLayoutArray.length;
+		/* The last element of the native signature array is for the return type */
+		String[] nativeSignatureStrs = new String[argLayoutCount + 1];
+		for (int argIndex = 0; argIndex < argLayoutCount; argIndex++) {
+			MemoryLayout argLayout = argLayoutArray[argIndex];
+			nativeSignatureStrs[argIndex] = LayoutStrPreprocessor.getSimplifiedLayoutString(argLayout, false);
+		}
+
+		/* Set the void layout string intended for the underlying native code
+		 * as the corresponding layout doesn't exist in the Spec.
+		 * Note: 'V' stands for the void type.
+		 */
+		if (realReturnLayout == null) {
+			nativeSignatureStrs[argLayoutCount] = "0#V"; //$NON-NLS-1$
+		} else {
+			nativeSignatureStrs[argLayoutCount] = LayoutStrPreprocessor.getSimplifiedLayoutString(realReturnLayout, false);
+		}
+
+		long addr = 0;
+		synchronized(privateUpcallClassLock) {
+			/* The generated thunks are shared across upcalls or threads only when
+			 * the target method handles are identical within the same session/scope
+			 * given a thunk plus the associated metadata is released automatically
+			 * in native via freeUpcallStub() in OpenJDK when the corresponding
+			 * session/scope is terminated.
+			 */
+			/*[IF JAVA_SPEC_VERSION >= 19]*/
+			String targetHashScopeStr = target.hashCode() + "#" + session.toString(); //$NON-NLS-1$
```

Comment:
why is this using `session.toString()`/`scope.toString()` instead of just hashcode?

Reply:
okay, what I meant was why not `target.hashCode() + "#" + session.hashCode()`? so hashcode for both

Reply:
The hash value of `session.hashCode()` is part of `session.toString()` as follows:
```
session.toString() = jdk.internal.foreign.ConfinedSession@5758ada0
session.hashCode() = 1465429408 <--- 5758ada0
```
More than that, `session.toString()` comes with more precise information of which type of session we are dealing with, which is very useful in troubleshooting in the future.